src/IO/PeekableReadBuffer.cpp

```diff
@@ -134,10 +143,11 @@ bool PeekableReadBuffer::nextImpl()
 
     checkStateCorrect();
     bool res;
+    bool checkpoint_at_end = checkpoint && *checkpoint == working_buffer.end();
 
     if (checkpoint)
     {
-        if (currentlyReadFromOwnMemory())
+        if (currentlyReadFromOwnMemory() || checkpoint_at_end)
```

Comment:
What is the purpose if this change? Seems like the case when checkpoint is at the end of working buffer is handled [here](https://github.com/ClickHouse/ClickHouse/blob/414f470c79eb22b0ca47b82f11625cf80b0231aa/src/IO/PeekableReadBuffer.cpp#L56).

Reply:
It's for the case when we put checkpoint on the edge of internal buffer and then call `next()`.

Reply:
BTW, it's really redundant here - removed.